source/optouts/OptOutsService.cs

```diff
@@ -0,0 +1,157 @@
+using System;
+using System.IO;
+using System.Net;
+using System.Text;
+using System.Web;
+using com.esendex.sdk.core;
+using com.esendex.sdk.exceptions;
+using com.esendex.sdk.optouts.models;
+using com.esendex.sdk.utilities;
+
+namespace com.esendex.sdk.optouts
+{
+    public class OptOutsService
+    {
+        private const string OPTOUTS_BASE_URL = "https://api.esendex.com";
+
+        private readonly Uri _baseUrl;
+        private readonly EsendexCredentials _credentials;
+
+        internal OptOutsService(string url, EsendexCredentials esendexCredentials)
+        {
+            _baseUrl = new Uri(url);
+            _credentials = esendexCredentials;
+        }
+
+        public OptOutsService(EsendexCredentials credentials) : this(OPTOUTS_BASE_URL, credentials)
+        {
+        }
+
+        public OptOutsService(string username, string password) : this(OPTOUTS_BASE_URL, new EsendexCredentials(username, password))
+        {
+        }
+
+        public OptOut GetById(Guid optOutId)
+        {
+            var requestUrl = new Uri(_baseUrl, string.Format("v1.0/optouts/{0}", optOutId));
+            var request = Request.Create("GET", requestUrl)
+                                 .WithHeader("Authorization", "Basic " + _credentials.EncodedValue())
+                                 .WithAcceptHeader(Constants.XML_MEDIA_TYPE)
+                                 .If(_credentials.UseProxy, r => r.WithProxy(_credentials.WebProxy));
+
+            try
+            {
+                var response = request.GetResponse();
+                using (var reader = new StreamReader(response.GetResponseStream(), Encoding.UTF8))
+                {
+                    string responseText = reader.ReadToEnd();
+                    return new XmlSerialiser().Deserialise<OptOut>(responseText);
+                }
+            }
+            catch (WebException ex)
+            {
+                var response = (HttpWebResponse)ex.Response;
+                if (response.StatusCode != HttpStatusCode.BadRequest)
+                    throw;
+
+                throw new BadRequestException(ex, null);
+            }
+        }
+
+        public SubscriptionCollection GetByPhoneNumber(string phoneNumber, string accountReference, int pageNumber, int pageSize)
+        {
+            var requestUrl = string.Concat(_baseUrl, "v1.0/optouts");
+            var uriBuilder = new UriBuilder(requestUrl);
+            var query = HttpUtility.ParseQueryString(string.Empty);
+            
+            if (accountReference != null)

```

Comment:
I did put it in a private method first, but it's only used in that one place and I thought it just creates more guff. I can understand breaking it out though because it makes it clear what is happening without having to read through it. Fair point